File: research/rsa4096-poc/pipeline.go
// This file wires steps 1-8 together into one callable pipeline, plus a
// self-contained "textbook RSA" correctness check that doesn't depend on any
// external library: encrypt/decrypt a test message with the raw (n, e, d)
// numbers and confirm it round-trips. This is independent of, and a
// necessary-but-not-sufficient complement to, the external Node/WebCrypto +
// real arweave-core cross-validation done by validate.mjs.
package main

import (
	"errors"
	"math/big"
)

// KeyGenResult bundles everything produced by one full run of the pipeline,
// for reporting and for writing out golden vectors (step 9).
type KeyGenResult struct {
	Seed        string
	P, Q        *big.Int
	Key         *RSAKey
	JWK         *JWK
	Address     string
	PAttempts   int
	QAttempts   int
}

// GenerateFullKey runs the complete seed -> RSA-4096 JWK pipeline (steps
// 1 through 6, plus the local half of step 8): open the seed stream, find
// two validated primes, assemble the key, encode the JWK, derive the
// address.
func GenerateFullKey(seedBitString string) (*KeyGenResult, error) {
	stream, err := NewSeedStream(seedBitString)
	if err != nil {
		return nil, err
	}

	p, q, pAttempts, qAttempts, err := FindTwoPrimes(stream)
	if err != nil {
		return nil, err
	}

	key, err := AssembleKey(p, q)
	if err != nil {
		return nil, err
	}

	jwk, err := key.ToJWK()
	if err != nil {
		return nil, err
	}

	address, err := AddressOf(key.N)
	if err != nil {
		return nil, err
	}

	return &KeyGenResult{
		Seed:      seedBitString,
		P:         p,
		Q:         q,
		Key:       key,
		JWK:       jwk,
		Address:   address,
		PAttempts: pAttempts,
		QAttempts: qAttempts,
	}, nil
}

// SelfCheckTextbookRSA verifies that d is genuinely the correct private
// exponent for (n, e) by round-tripping several test messages through raw
// (unpadded) RSA: c = m^e mod n, m' = c^d mod n, and checking m' == m. This
// is deliberately independent of any padding scheme (RSA-PSS, PKCS#1 v1.5,
// etc.) -- it isolates and proves the one thing that could actually be
// wrong at this stage: whether AssembleKey computed a mathematically
// correct d. Padding-scheme correctness is exercised separately by
// validate.mjs against real WebCrypto RSA-PSS.
func SelfCheckTextbookRSA(key *RSAKey) error {
	testMessages := []int64{2, 3, 1234567, 987654321}
	for _, mVal := range testMessages {
		m := big.NewInt(mVal)
		if m.Cmp(key.N) >= 0 {
			return errors.New("SelfCheckTextbookRSA: test message >= n, test is malformed")
		}

		c := new(big.Int).Exp(m, key.E, key.N)   // encrypt: c = m^e mod n
		mPrime := new(big.Int).Exp(c, key.D, key.N) // decrypt: m' = c^d mod n

		if m.Cmp(mPrime) != 0 {
			return errors.New("SelfCheckTextbookRSA: round-trip FAILED for message " + m.String() + " -- d is not the correct inverse of e")
		}

		// Also confirm signing direction (sign with d, verify with e) --
		// for RSA these are the same operation mathematically (m^d mod n
		// then result^e mod n), but worth checking explicitly since that's
		// the operation actually used for Arweave transaction signing.
		s := new(big.Int).Exp(m, key.D, key.N)
		mFromSig := new(big.Int).Exp(s, key.E, key.N)
		if m.Cmp(mFromSig) != 0 {
			return errors.New("SelfCheckTextbookRSA: sign/verify round-trip FAILED for message " + m.String())
		}
	}
	return nil
}

File: ts/src/rsa4096/pipeline.ts
/**
 * Wires the seed -> RSA-4096 JWK pipeline together, plus a self-contained
 * "textbook RSA" correctness check independent of any external library.
 * Direct port of RSA4096/pipeline.go.
 *
 * Copyright (C) 2026 AncientHoldings GmbH. All rights reserved.
 */

import { modPow } from './bigint-math.js';
import { type JWK, addressOf, toJWK } from './jwk.js';
import { type RSAKey, assembleKey } from './keyassembly.js';
import { findTwoPrimes, findTwoPrimesAsync } from './primesearch.js';
import type { ProgressCallback } from './progress.js';
import { newSeedStream } from './stream.js';

/** Bundles everything produced by one full run of the pipeline. */
export interface KeyGenResult {
  seed: string;
  p: bigint;
  q: bigint;
  key: RSAKey;
  jwk: JWK;
  address: string;
  pAttempts: number;
  qAttempts: number;
}

/**
 * Runs the complete seed -> RSA-4096 JWK pipeline: open the seed stream,
 * find two validated primes, assemble the key, encode the JWK, derive the
 * address. Matches RSA4096/pipeline.go's GenerateFromBitString exactly.
 *
 * `onProgress` is optional and purely observational (see progress.ts) --
 * a caller building a UI progress bar passes a callback here. Every
 * existing test and tool that omits it gets byte-for-byte identical
 * output to before progress reporting existed.
 *
 * This is the SYNCHRONOUS variant: it blocks the calling thread for the
 * full multi-second search. In a browser, that means the UI thread (and
 * therefore any progress bar built from `onProgress`) will NOT repaint
 * during the search -- use {@link generateFromBitStringAsync} instead for
 * an actual live-updating browser progress bar. This variant remains the
 * right default for Node/server contexts where blocking is acceptable
 * (same tradeoff already documented for `scalarMultiplier` vs
 * `scalarMultiplierAsync` in gen1/scalar-mult.ts).
 */
export function generateFromBitString(
  seedBitString: string,
  onProgress?: ProgressCallback,
): KeyGenResult {
  const stream = newSeedStream(seedBitString);
  const { p, q, pAttempts, qAttempts } = findTwoPrimes(stream, onProgress);
  const key = assembleKey(p, q);
  const jwk = toJWK(key);
  const address = addressOf(key.n);

  return { seed: seedBitString, p, q, key, jwk, address, pAttempts, qAttempts };
}

/**
 * Async variant of {@link generateFromBitString}, built on
 * {@link findTwoPrimesAsync}: yields to the event loop every 8 candidate
 * draws (see primesearch.ts), so a browser UI thread stays responsive and
 * an `onProgress` callback can actually drive a repainting progress bar
 * for the full duration of the search -- this is the recommended entry
 * point for any UI use of this package.
 */
export async function generateFromBitStringAsync(
  seedBitString: string,
  onProgress?: ProgressCallback,
): Promise<KeyGenResult> {
  const stream = newSeedStream(seedBitString);
  const { p, q, pAttempts, qAttempts } = await findTwoPrimesAsync(stream, onProgress);
  const key = assembleKey(p, q);
  const jwk = toJWK(key);
  const address = addressOf(key.n);

  return { seed: seedBitString, p, q, key, jwk, address, pAttempts, qAttempts };
}

/**
 * Verifies that `d` is genuinely the correct private exponent for (n, e)
 * by round-tripping several test messages through raw (unpadded) RSA:
 * c = m^e mod n, m' = c^d mod n, and checking m' === m -- deliberately
 * independent of any padding scheme (RSA-PSS, PKCS#1 v1.5, etc.). Matches
 * RSA4096/pipeline.go's SelfCheckTextbookRSA exactly.
 */
export function selfCheckTextbookRSA(key: RSAKey): void {
  const testMessages = [2n, 3n, 1234567n, 987654321n];
  for (const m of testMessages) {
    if (m >= key.n) {
      throw new Error('selfCheckTextbookRSA: test message >= n, test is malformed');
    }

    const c = modPow(m, key.e, key.n); // encrypt: c = m^e mod n
    const mPrime = modPow(c, key.d, key.n); // decrypt: m' = c^d mod n
    if (mPrime !== m) {
      throw new Error(
        `selfCheckTextbookRSA: round-trip FAILED for message ${m} -- d is not the correct inverse of e`,
      );
    }

    // Sign/verify direction (sign with d, verify with e) -- for RSA these
    // are the same operation mathematically, but worth checking
    // explicitly since it's the operation actually used for Arweave
    // transaction signing.
    const s = modPow(m, key.d, key.n);
    const mFromSig = modPow(s, key.e, key.n);
    if (mFromSig !== m) {
      throw new Error(`selfCheckTextbookRSA: sign/verify round-trip FAILED for message ${m}`);
    }
  }
}
